File: problems_list/0200.岛屿数量/0200-岛屿数量.js
/**
 * @param {character[][]} grid
 * @return {number}
 */
var numIslands = function (grid) {
  let rows = grid.length,
    cols = grid[0].length;
  // 方向数组
  let directions = [
    [0, 1],
    [0, -1],
    [-1, 0],
    [1, 0],
  ];
  // 深度优先搜索
  const dfs = (i, j) => {
    // 超出边界 或者本身就已经是海水了
    if (i < 0 || j < 0 || i >= rows || j >= cols || grid[i][j] == 0) return;
    // 淹没它 避免重复访问
    grid[i][j] = 0;
    for (let dir of directions) dfs(dir[0] + i, dir[1] + j);
  };
  let count = 0;
  for (let i = 0; i < rows; i++) {
    for (let j = 0; j < cols; j++) {
      if (grid[i][j] == 1) {
        count++;
        dfs(i, j);
      }
    }
  }
  return count;
};

File: problems_list/0200.岛屿数量/0200-岛屿数量.cpp
class Solution {
public:
    int dirs[4][2]={{-1,0},{0,1},{1,0},{0,-1}};
    int m,n;
    void dfs(vector<vector<char>>& matrix, int x, int y){
        matrix[x][y]='#';
        int nx=-1, ny=-1;
        for(int i=0; i<4; i++){
            nx=x+dirs[i][0], ny=y+dirs[i][1];
            if(nx>=0 && nx<m && ny>=0 && ny<n && matrix[nx][ny]=='1'){
                dfs(matrix, nx, ny);
            }
        }
    }
    int numIslands(vector<vector<char>>& grid) {
        int m=grid.size(), n=grid[0].size();
        this->m=m, this->n=n;
        int ans=0;
        for(int i=0; i<m; i++){
            for(int j=0; j<n; j++){
                if(grid[i][j]=='1'){
                    dfs(grid, i, j);
                    ans++;
                }
            }
        }
        return ans;
    }
};